lib/Infer/AbstractInterpreter.cpp

```diff
@@ -1,6 +1,272 @@
 #include "souper/Infer/Interpreter.h"
+#include "souper/Infer/AbstractInterpreter.h"
+
+#include <iostream>
+
+using namespace llvm;
+
+namespace {
+
+  APInt getUMin(const KnownBits &x) { return x.One; }
+
+  APInt getUMax(const KnownBits &x) { return ~x.Zero; }
+
+  bool isSignKnown(const KnownBits &x) {
+    unsigned W = x.getBitWidth();
+    return x.One[W - 1] || x.Zero[W - 1];
+  }
+
+  APInt getSMin(const KnownBits &x) {
+    if (isSignKnown(x))
+      return x.One;
+    APInt Min = x.One;
+    Min.setBit(x.getBitWidth() - 1);
+    return Min;
+  }
+
+  APInt getSMax(const KnownBits &x) {
+    if (isSignKnown(x))
+      return ~x.Zero;
+    APInt Max = ~x.Zero;
+    Max.clearBit(x.getBitWidth() - 1);
+    return Max;
+  }
+} // anonymous
 
 namespace souper {
+
+  namespace BinaryTransferFunctionsKB {
+    llvm::KnownBits add(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/true, /*NSW=*/false,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits addnsw(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/true, /*NSW=*/true,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits sub(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/false, /*NSW=*/false,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits subnsw(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      return llvm::KnownBits::computeForAddSub(/*Add=*/false, /*NSW=*/true,
+                                               lhs, rhs);
+    }
+
+    llvm::KnownBits mul(const llvm::KnownBits &lhs, const llvm::KnownBits &rhs) {
+      llvm::KnownBits Result(lhs.getBitWidth());
+
+      auto trailingZeros0 = lhs.countMinTrailingZeros();
+      auto trailingZeros1 = rhs.countMinTrailingZeros();
+      Result.Zero.setLowBits(std::min(trailingZeros0 + trailingZeros1, lhs.getBitWidth()));
+
+      // check for leading zeros
```

Comment:
I think there are good opportunties for improvement here by counting leading ones as well as zeroes, right? no need to fix it now but please add a comment sketching out the necessary code